### PR TITLE
HTTP/3: Disable HTTP/3 when multiple protocols used on dynamic port

### LINKED ISSUE #42982: HTTP/3: Ephemeral port and connection already in use
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

A set of ASP.NET Core HTTP/3 tests failed - https://github.com/dotnet/aspnetcore/issues/42967. They are both configured to start on 127.0.0.1:0. Because the port is 0 (ephemeral) the listener should start on an unused port assigned by Windows.

Instead, this error is thrown: `System.Net.Quic.QuicException : The local address is already in use.`

The problem is Kestrel listening on HTTP/1.1 (TCP), HTTP/2 (TCP), and HTTP/3 (UDP).

The ephemeral port is used to bind a TCP socket first, which returns the assigned port. Then that port is reused with QUIC because we want the listener connections to share the same port number.

The problem is there's a chance that the free port for TCP is in use with UDP, causing this error.

### Expected Behavior

1. Kestrel starts up and is free from address in use errors.
2. The endpoint shares the same port numbers.

### Steps To Reproduce

Kestrel + epemeral port + HTTP/1.1, HTTP/2, HTTP/3

The error will only happen if a TCP port is assigned and the same number UDP port is in use.

### Exceptions (if any)

`System.Net.Quic.QuicException : The local address is already in use.`

### .NET Version

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/42982

### CODE REVIEW

File: src/Servers/Kestrel/Core/test/KestrelServerTests.cs [hunk 1]
```diff
@@ -318,7 +318,7 @@ public async Task ListenIPWithEphemeralPort_TransportsGetIPv6Any()
         options.ApplicationServices = new ServiceCollection()
            .AddLogging()
            .BuildServiceProvider();
-        options.ListenAnyIP(0, options =>
+        options.ListenAnyIP(5000, options =>
```
Comment (halter73) [inline src/Servers/Kestrel/Core/test/KestrelServerTests.cs:321]: Given the name `ListenIPWithEphemeralPort_TransportsGetIPv6Any`, I think we should update this test to disable HTTP/3 instead.

File: src/Servers/Kestrel/Transport.Quic/test/WebHostTests.cs [hunk 2]
```diff
@@ -122,7 +142,7 @@ public async Task Listen_Http3AndSocketsCoexistOnSameEndpoint_ClientSuccess()
                 webHostBuilder
                     .UseKestrel(o =>
                     {
-                        o.Listen(IPAddress.Parse("127.0.0.1"), 5005, listenOptions =>
+                        o.Listen(IPAddress.Parse("127.0.0.1"), 0, listenOptions =>
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Quic/test/WebHostTests.cs:127]: Nit: I know other tests make use of this, but should we have a test specifically verifying port-0 binding works when just HTTP/3 is enabled?

File: src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs [hunk 3]
```diff
@@ -211,18 +213,27 @@ async Task OnBind(ListenOptions options, CancellationToken onBindCancellationTok
...
+                    options.EndPoint = await _transportManager.BindAsync(configuredEndpoint, connectionDelegate, options.EndpointConfig, onBindCancellationToken).ConfigureAwait(false);
                 }
 
                 if (hasHttp3 && _multiplexedTransportFactory is not null)
                 {
-                    options.UseHttp3Server(ServiceContext, application, options.Protocols, addAltSvcHeader);
-                    var multiplexedConnectionDelegate = ((IMultiplexedConnectionBuilder)options).Build();
+                    // Check if a previous transport has changed the endpoint. If it has then the endpoint is dynamic and we can't guarantee it will work for other transports.
+                    // For more details, see https://github.com/dotnet/aspnetcore/issues/42982
+                    if (!configuredEndpoint.Equals(options.EndPoint))
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs:223]: I probably would have just tried casting to an IPEndpoint and checked for port-0 up front, but I like this way of checking for dynamic binding.

**Review by halter73 — APPROVED** [review]
I want `ListenIPWithEphemeralPort_TransportsGetIPv6Any` updated to use an ephemeral port again though.

File: src/Servers/Kestrel/Transport.Quic/test/WebHostTests.cs (hunk 2 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Transport.Quic/test/WebHostTests.cs:127]: I don't think it's necessary. Lots of tests rely on it working.

File: src/Servers/Kestrel/Core/test/KestrelServerTests.cs (hunk 1 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/test/KestrelServerTests.cs:321]: Duplicated this test but for HTTP/3 plus multiplex.

**Pull request merged**